readme.txt
nes_ntsc 0.2.3: NES NTSC Video Filter
-------------------------------------
This library filters a NES image to match what a TV would show, allowing
an authentic image in an emulator. It uses a highly optimized algorithm
to perform the same signal processing as an NTSC decoder in a TV, giving
very similar pixel artifacts and color bleeding. The usual picture
controls can be adjusted: hue, saturation, contrast, brightness, and
sharpness. Additionally, the amount of NTSC chroma and luma artifacts
can be reduced, allowing an image that corresponds to composite video
(artifacts), S-video (color bleeding only), RGB (clean pixels), or
anywhere inbetween.

The output is scaled to the proper horizontal width, leaving it up the
emulator to simply double the height. An optional even/odd field merging
feature is provided to reduce flicker when the host display's refresh
rate isn't 60 Hz. Specialized blitters can be easily written using a
special interface, allowing customization of input and output pixel
formats, optimization for the host platform, and efficient scanline
doubling.

Blitting a 256x240 source image to a 602x240 pixel 16-bit RGB memory
buffer at 60 frames per second uses 8% CPU on a 2.0 GHz Athlon 3500+ and
40% CPU on a 10-year-old 400 MHz G3 PowerMac.

Author  : Shay Green <gblargg@gmail.com>
Website : http://www.slack.net/~ant/
Forum   : http://groups.google.com/group/blargg-sound-libs
License : GNU Lesser General Public License (LGPL)
Language: C or C++


Getting Started
---------------
Build a program from demo.c, nes_ntsc.c, and the SDL multimedia library
(see http://libsdl.org/). Run it with "test.bmp" in the same directory
and it should show the filtered image. See demo.c for more.

See nes_ntsc.txt for documentation and nes_ntsc.h for reference. Post to
the discussion forum for assistance.


Files
-----
readme.txt          Essential information
nes_ntsc.txt        Library documentation
changes.txt         Changes made since previous releases
license.txt         GNU Lesser General Public License

benchmark.c         Measures frame rate and processor usage of library
demo.c              Displays and saves NTSC filtered image
demo_impl.h         Internal routines used by demo
test.bmp            Test image for demo
tests/              Test ROMs to verify burst phase operation

nes_ntsc_config.h   Library configuration (modify as needed)
nes_ntsc.h          Library header and source
nes_ntsc.c
nes_ntsc_impl.h

-- 
Shay Green <gblargg@gmail.com>


changes.txt
nes_ntsc Change Log
-------------------

nes_ntsc 0.2.4
-------------
- Added commandline arguments and flags, see blarg_ntsc.txt for more info

nes_ntsc 0.2.3
--------------
- Updated SDL dependency to 2.0; updated SDL code in demo_impl.h accordingly
- Added toggleable scanline rendering

nes_ntsc 0.2.2
--------------
- Added ability to use a custom RGB palette in place of the NES color
circuitry (with the option of being before color emphasis). When using
this palette with otherwise default settings, RGB colors appear on
screen exactly as in custom palette.

- Moved configuration options to nes_ntsc_config.h, making it easier to
manage

- Made color emphasis support configurable in nes_ntsc_config.h (off by
default). Previously, there were separate functions; now there is only
one set of functions and whether they support color emphasis is
configurable. If color emphasis support is OFF, then nes_ntsc_init() and
related behave as before, and the nes_ntsc_init_emph() functionality is
not available. If color emphasis support is ON, then nes_ntsc_init() and
related behave as nes_ntsc_init_emph() and related did before, and
nes_ntsc_init() and related functionality is not available (so you would
need to update any code that uses palette_out and expects only 64 colors
to be written; see the end of demo.c for example).

- Removed hue_warping for now, due to its obscurity

- Greatly clarified and improved demo to read any uncompressed BMP image
and write filtered image when done

- Improved gamma to be properly applied to each RGB channel, and changed
default to compensate for difference between PC monitor and TV gamma

- Improved contrast to be properly applied to each RGB channel rather
than just luma

- Improved floating point calculations in library to be more stable and
not need double precision, which was causing problems with the sharpness
control on Windows when the DirectX libraries changed the FPU to single
precision mode

- Eliminated slight artifacts still visible when using the RGB preset
with field merging off

- Added extern "C" to header, allowing use in C++ without having to
rename the source file

- Made internal changes to factor out code common from all my NTSC
filter libraries, greatly simplifying things for me


nes_ntsc 0.2.0
--------------
- Significantly improved NTSC signal processing to give clearer image
and better sharpness control

- Added parameters for gamma, resolution, color bleed, artifacts, and
color fringing

- Added presets for composite video, S-video, RGB, and monochrome

- Added NES_NTSC_OUT_WIDTH() and NES_NTSC_IN_WIDTH() for calculating
input/output widths

- Improved default blitter to support emphasis and allow specification
of RGB output bit depth

- Improved demo with more controls and interpolation and darkening of
scanlines rather than duplicating them

- Improved documentation

- Interface changes: nes_ntsc_blit() now takes *input* size rather than
output size. NES_NTSC_BEGIN_ROW now takes two additional pixels.

- Deprecated: nes_ntsc_x_in/out_width, NES_NTSC_RGBnn_OUT, and
NES_NTSC_RAW_OUT


nes_ntsc 0.1.7
--------------
- Fixed color emphasis to affect xD colors instead of incorrectly
leaving them as always gray

- Added ability to generate a 64- or 512-color RGB palette for use when
full NTSC emulation isn't desired

- Documented more accurate burst_phase handling and included three test
ROMs for verifying proper implementation

- Moved color emphasis support to new nes_ntsc_emph_t structure,
eliminating the ugly global NES_NTSC_DISABLE_EMPHASIS configuration
macro

- Extended demo to use optional Sony decoder matrix and write standard
192-byte NES .PAL file using current settings

- Improved documentation slightly

- Lowered maximum saturation to avoid overflow of blue when blue color
emphasis is enabled


nes_ntsc 0.1.6
--------------
- Cleaned up hue, decoder matrix, and hue warping to do transformations
in the proper order and direction


nes_ntsc 0.1.5
--------------
- Added support for color emphasis/tint bits (can be turned off to
reduce memory usage)

- Added ability to specify optional decoder matrix

- Improved documentation

- Added constants for overscan borders


nes_ntsc 0.1.0
--------------
- First official version

- Completely rewrote algorithm to do all NTSC signal processing at
initialization time, putting results into a table for use during
blitting. This increased performance by over 300% and allowed many
improvements in quality and image options.

- Eliminated multiple output formats and options. Now you can define a
custom blitter for this.

- Added sharpness and "hue warping" controls

- Added option to merge even and odd artifacts to reduce flicker when
host monitor's refresh rate doesn't match emulator's frame rate.

- Added built-in horizontal rescaling. I never realized the old one was
way too wide.


demo_impl.h
/* Simple shell used by demos. Uses SDL multimedia library. */

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include "SDL.h"

/* Image loader */
typedef struct image_t
{
	unsigned char const* byte_pixels;/* 8-bit pixels */
	unsigned short const* rgb_16;    /* 16-bit pixels */
	int width;
	int height;
	int row_width; /* number of pixels to get to next row (may be greater than width) */
} image_t;
/* if no palette, loads as 16-bit RGB */
void load_bmp( image_t* out, const char* path, SDL_Color palette [256] );
void save_bmp( const char* path );
void init_window( int width, int height );
int read_input( void );
void lock_pixels( void );
void double_output_height( void );
void display_output( void );
void fatal_error( const char* str );

static unsigned char* output_pixels; /* 16-bit RGB */
static long output_pitch;
static float mouse_x, mouse_y; /* -1.0 to 1.0 */
static int mouse_moved;
static int key_pressed;
static int scanlines = 1;

/* implementation */

static SDL_Window* screen;
static SDL_Surface* surface;
static SDL_Texture* texture;
static SDL_Renderer* renderer;
static unsigned long next_time;

void fatal_error( const char* str )
{
	fprintf( stderr, "Error: %s\n", str );
	exit( EXIT_FAILURE );
}

void cmd_usage(void)
{
	fprintf(stdout, "usage:\nnes_ntsc\t[input.bmp] [output.bmp]\n");
	fprintf(stdout, "nes_ntsc\tinput.bmp output.bmp -c [-scanoff] [-mergeoff <burstphase>] [-sharpness <value>] [-gamma <value>] [-vidmode cvbs|svid|rgb|mono] [-sony]\n\n");
	fprintf(stdout, "\t[input.bmp]\tInput.bmp file. \"test.bmp\" by default.\n");
	fprintf(stdout, "\t[output.bmp]\tOutput .bmp file. \"filtered.bmp\" by default.\n");
	fprintf(stdout, "\t-c\t\tUses the following commandline parameters instead of opening a window:\n");
	fprintf(stdout, "\t[-scanoff]\tTurns off scanline effects. On by default.\n");
	fprintf(stdout, "\t[-mergeoff]\tDisables merging fields. Burst phase is 1 by default.\n");
	fprintf(stdout, "\t[-sharpness]\t<value> ranges from -1.00 to 1.00. 0 by default.\n");
	fprintf(stdout, "\t[-gamma]\t<value> ranges from -1.00 to 1.00. 0 by default.\n");
	fprintf(stdout, "\t[-vidmode]\tSelects a video mode. \"cvbs\" by default.\n");
	fprintf(stdout, "\t[-sony]\t\tTurns on Sony decoder. Off by default.\n");
}

static void init_sdl_( void )
{
	static int initialized;
	if ( !initialized )
	{
		if ( SDL_Init( SDL_INIT_VIDEO ) < 0 )
			fatal_error( "SDL initialization failed" );
		atexit( SDL_Quit );
	}
}

void init_window( int width, int height )
{
	init_sdl_();

	screen = SDL_CreateWindow("NTSC Filter Demo",
		SDL_WINDOWPOS_UNDEFINED,
		SDL_WINDOWPOS_UNDEFINED,
		width, height,
		0);
	renderer = SDL_CreateRenderer(screen, -1, 0);
	SDL_SetRenderDrawColor(renderer, 255, 0, 255, 255); // debug
	surface = SDL_CreateRGBSurface(0, width, height, 16, 0, 0, 0, 0);
	texture = SDL_CreateTexture(renderer,
		SDL_PIXELFORMAT_RGB565,
		SDL_TEXTUREACCESS_STREAMING,
		width, height);
	;
	if ( screen == NULL || renderer == NULL || surface == NULL || texture == NULL )
		fatal_error( "SDL initialization failed" );
}

int read_input( void )
{
	SDL_Event e;
	
	/* limit to 60 calls per second */
	unsigned long start = SDL_GetTicks();
	if ( start < next_time && next_time - start > 10 )
		SDL_Delay( next_time - start );
	while ( SDL_GetTicks() < next_time ) { }
	next_time = start + 1000 / 60;
	
	mouse_moved = 0;
	key_pressed = 0;
	
	while ( SDL_PollEvent( &e ) )
	{
		if ( e.type == SDL_MOUSEBUTTONDOWN || e.type == SDL_QUIT )
			return 0;
		
		if ( e.type == SDL_KEYDOWN )
		{
			if ( e.key.keysym.sym == SDLK_ESCAPE || e.key.keysym.sym == SDLK_q )
				return 0;
			key_pressed = e.key.keysym.sym;
		}
		
		if ( e.type == SDL_MOUSEMOTION )
		{
			int x, y;
			SDL_GetMouseState( &x, &y );
			mouse_moved = 1;
			mouse_x = x / (float) (SDL_GetWindowSurface(screen)->w - 1) * 2 - 1;
			mouse_y = (1 - y / (float) (SDL_GetWindowSurface(screen)->h - 1)) * 2 - 1;
		}
	}
	return 1;
}

void lock_pixels( void )
{
	if ( SDL_LockSurface( surface ) < 0 )
		fatal_error( "Couldn't lock surface" );
	SDL_FillRect( surface, 0, 0 );
	output_pitch = surface->pitch;
	output_pixels = (unsigned char*) surface->pixels;
}

void double_output_height( void )
{
	int y;
	for ( y = surface->h / 2; --y >= 0; )
	{
		unsigned char const* in = output_pixels + y * output_pitch;
		unsigned char* out = output_pixels + y * 2 * output_pitch;
		int n;
		for ( n = surface->w; n; --n )
		{
			unsigned prev = *(unsigned short*) in;
			unsigned next = *(unsigned short*) (in + output_pitch);
			/* mix 16-bit rgb without losing low bits */
			unsigned mixed = prev + next + ((prev ^ next) & 0x0821);
			/* darken by 12% */
			*(unsigned short*) out = prev;
			if (scanlines)
				*(unsigned short*) (out + output_pitch) = (mixed >> 1) - (mixed >> 4 & 0x18E3);
			else
				*(unsigned short*)(out + output_pitch) = prev;
			in += 2;
			out += 2;
		}
	}
}

void display_output( void )
{
	SDL_UnlockSurface( surface );
	if (SDL_UpdateTexture(texture, NULL, surface->pixels, surface->pitch) < 0
		|| SDL_RenderClear(renderer) < 0
		|| SDL_RenderCopy(renderer, texture, NULL, NULL) < 0)
		fatal_error("SDL renderer failed");
	SDL_RenderPresent(renderer);

}

void load_bmp( image_t* out, const char* path, SDL_Color palette [256] )
{	
	SDL_PixelFormat fmt = { 0 }; /* clear fields */
	SDL_Palette pal = { 0 };
	SDL_Surface* bmp;
	SDL_Surface* conv;
	
	init_sdl_();
	bmp = SDL_LoadBMP( path );
	if ( !bmp )
		fatal_error( "Couldn't load BMP" );
	
	fmt.BitsPerPixel  = 16;
	fmt.BytesPerPixel = 2;
	if ( palette )
	{
		pal.ncolors = 256;
		pal.colors = palette;
		fmt.palette = &pal;
		fmt.BitsPerPixel  = 8;
		fmt.BytesPerPixel = 1;
	}
	conv = SDL_ConvertSurface( bmp, &fmt, SDL_SWSURFACE );
	if ( !conv )
		fatal_error( "Couldn't convert BMP" );
	SDL_FreeSurface( bmp );
	
	if ( SDL_LockSurface( conv ) < 0 )
		fatal_error( "Couldn't lock surface" );
	
	out->byte_pixels = (unsigned char *) conv->pixels;
	out->rgb_16      = (unsigned short*) conv->pixels;
	out->width       = conv->w;
	out->height      = conv->h;
	out->row_width   = conv->pitch / fmt.BytesPerPixel;
}

void save_bmp( const char* path )
{
	if ( SDL_SaveBMP( surface, path ) )
		fatal_error( "Couldn't save BMP" );
}


demo.c
/* Displays and saves NTSC filtered image. Mouse controls sharpness and gamma.
Defaults to using "test.bmp" for input and "filtered.bmp" for output. Input
image must be an uncompressed BMP. Also writes "nes.pal" RGB color file on exit.

Usage: demo [in.bmp [out.bmp]]

Space   Toggle field merging
C       Composite video quality
S       S-video quality
R       RGB video quality
M       Monochrome video quality
D       Toggle between standard and Sony decoder matrix
*/

#include "nes_ntsc.h"

#include "demo_impl.h"

/* only used to convert input image to native palette format */
static SDL_Color palette [256] = {
	{102,102,102},{  0, 42,136},{ 20, 18,168},{ 59,  0,164},
	{ 92,  0,126},{110,  0, 64},{108,  7,  0},{ 87, 29,  0},
	{ 52, 53,  0},{ 12, 73,  0},{  0, 82,  0},{  0, 79,  8},
	{  0, 64, 78},{  0,  0,  0},{  0,  0,  0},{  0,  0,  0},
	{174,174,174},{ 21, 95,218},{ 66, 64,254},{118, 39,255},
	{161, 27,205},{184, 30,124},{181, 50, 32},{153, 79,  0},
	{108,110,  0},{ 56,135,  0},{ 13,148,  0},{  0,144, 50},
	{  0,124,142},{  0,  0,  0},{  0,  0,  0},{  0,  0,  0},
	{254,254,254},{100,176,254},{147,144,254},{199,119,254},
	{243,106,254},{254,110,205},{254,130,112},{235,159, 35},
	{189,191,  0},{137,217,  0},{ 93,229, 48},{ 69,225,130},
	{ 72,206,223},{ 79, 79, 79},{  0,  0,  0},{  0,  0,  0},
	{254,254,254},{193,224,254},{212,211,254},{233,200,254},
	{251,195,254},{254,197,235},{254,205,198},{247,217,166},
	{229,230,149},{208,240,151},{190,245,171},{180,243,205},
	{181,236,243},{184,184,184},{  0,  0,  0},{  0,  0,  0}
};

int main(int argc, char** argv)
{
	if (argc > 1 && strcmp(argv[1], "-h") == 0)
		cmd_usage();
	else
	{
	image_t image;
	int sony_decoder = 0;
	int merge_fields = 1;
	int burst_phase = 0;
	nes_ntsc_setup_t setup = nes_ntsc_composite;

	nes_ntsc_t* ntsc = (nes_ntsc_t*)malloc(sizeof(nes_ntsc_t));
	if (!ntsc)
		fatal_error("Out of memory");
	nes_ntsc_init(ntsc, &setup);

		load_bmp(&image, (argc > 1 ? argv[1] : "test.bmp"), palette);
		init_window(NES_NTSC_OUT_WIDTH(image.width), image.height * 2);

		if (argc > 2 && strcmp(argv[3], "-c") == 0)
		{

			// handle commandline flags
			burst_phase = 1;
			float sharpness = 0.0;
			float gamma = 0.0;
			char videomode[] = "NTSC composite";
			int i;
			for (i = 0; i < argc; i++)
			{
				if (strcmp(argv[i], "-scanoff") == 0)
					scanlines = 0;

				if (strcmp(argv[i], "-mergeoff") == 0) {
					merge_fields = 0;
					burst_phase = atoll(argv[i + 1]);
				}

				if (strcmp(argv[i], "-sharpness") == 0)
					sharpness = strtof(argv[i + 1], NULL);

				if (strcmp(argv[i], "-gamma") == 0)
					gamma = strtof(argv[i + 1], NULL);

				// how do I switch case with strings
				if (strcmp(argv[i], "-vidmode") == 0) {
					if (strcmp(argv[i + 1], "svid") == 0) {
						setup = nes_ntsc_svideo;
						strcpy(videomode, "S-Video");
					}
					else if (strcmp(argv[i + 1], "rgb") == 0) {
						setup = nes_ntsc_rgb;
						strcpy(videomode, "RGB");
					}
					else if (strcmp(argv[i + 1], "mono") == 0) {
						setup = nes_ntsc_monochrome;
						strcpy(videomode, "Monochrome");
					}
					else{
						strcpy(videomode, "NTSC composite");
					}
				}

				if (strcmp(argv[i], "-sony") == 0)
					sony_decoder = 1;
			}
			setup.merge_fields = merge_fields;
			setup.sharpness = sharpness;
			setup.gamma = gamma;

			setup.decoder_matrix = 0;
			if (sony_decoder)
			{
				/* Sony CXA2025AS US */
				static float matrix[6] = { 1.630, 0.317, -0.378, -0.466, -1.089, 1.677 };
				setup.decoder_matrix = matrix;
			}
			// I have no idea why I need to call nes_ntsc_init() twice
			nes_ntsc_init(ntsc, &setup);
			
			lock_pixels();

			if (setup.merge_fields)
				burst_phase = 0;

			nes_ntsc_blit(ntsc, image.byte_pixels, image.row_width, burst_phase,
				image.width, image.height, output_pixels, output_pitch);

			double_output_height();
			display_output();

			nes_ntsc_init(ntsc, &setup);

			printf("\tscanline enabled:\t%i\n", scanlines);
			printf("\tmerge fields enabled:\t%i\n", merge_fields);
			printf("\tsharpness:\t\t%f\n", sharpness);
			printf("\tgamma:\t\t\t%f\n", gamma);
			printf("\tvideo mode:\t\t%s\n", videomode);
			printf("\tsony decoder enabled:\t%i\n", sony_decoder);
		}
		else
		{
			while (read_input())
			{
				lock_pixels();

				burst_phase ^= 1;
				if (setup.merge_fields)
					burst_phase = 0;

				nes_ntsc_blit(ntsc, image.byte_pixels, image.row_width, burst_phase,
					image.width, image.height, output_pixels, output_pitch);

				double_output_height();
				display_output();

				switch (key_pressed)
				{
				case ' ': merge_fields = !merge_fields; break;
				case 'a': scanlines = !scanlines; break;
				case 'c': setup = nes_ntsc_composite; break;
				case 's': setup = nes_ntsc_svideo; break;
				case 'r': setup = nes_ntsc_rgb; break;
				case 'm': setup = nes_ntsc_monochrome; break;
				case 'd': sony_decoder = !sony_decoder; break;
				}

				if (key_pressed || mouse_moved)
				{
					setup.merge_fields = merge_fields;

					/* available parameters: hue, saturation, contrast, brightness,
					sharpness, gamma, bleed, resolution, artifacts, fringing */
					setup.sharpness = mouse_x;
					setup.gamma = mouse_y;

					setup.decoder_matrix = 0;
					if (sony_decoder)
					{
						/* Sony CXA2025AS US */
						static float matrix[6] = { 1.630, 0.317, -0.378, -0.466, -1.089, 1.677 };
						setup.decoder_matrix = matrix;
					}

					nes_ntsc_init(ntsc, &setup);
					fprintf(stdout, " sharpness: %.2f, gamma: %.2f \r", mouse_x, mouse_y);
				}
			}
		}
		save_bmp(argc > 2 ? argv[2] : "filtered.bmp");
		free(ntsc);

		/* write standard 192-byte NES palette */
		{
			FILE* out = fopen("nes.pal", "wb");
			if (out)
			{
				unsigned char palette[nes_ntsc_palette_size * 3];
				setup.palette_out = palette;
				nes_ntsc_init(0, &setup);
				fwrite(palette, 192, 1, out);
			}
		}
	}
	return 0;
}